File: scripts/codereview/ts/ast-extractor.ts
import * as ts from 'typescript';
import * as fs from 'fs';
import * as path from 'path';

interface Param {
  name: string;
  type: string;
}

interface FuncSig {
  params: Param[];
  returns: string[];
  is_async: boolean;
  is_exported: boolean;
  start_line: number;
  end_line: number;
}

interface FunctionDiff {
  name: string;
  change_type: 'added' | 'removed' | 'modified' | 'renamed';
  before?: FuncSig;
  after?: FuncSig;
  body_diff?: string;
}

interface FieldDiff {
  name: string;
  change_type: 'added' | 'removed' | 'modified';
  old_type?: string;
  new_type?: string;
}

interface TypeDiff {
  name: string;
  kind: string;
  change_type: 'added' | 'removed' | 'modified' | 'renamed';
  fields?: FieldDiff[];
  start_line: number;
  end_line: number;
}

interface ImportDiff {
  path: string;
  alias?: string;
  change_type: 'added' | 'removed';
}

interface ChangeSummary {
  functions_added: number;
  functions_removed: number;
  functions_modified: number;
  types_added: number;
  types_removed: number;
  types_modified: number;
  imports_added: number;
  imports_removed: number;
}

interface SemanticDiff {
  language: string;
  file_path: string;
  functions: FunctionDiff[];
  types: TypeDiff[];
  imports: ImportDiff[];
  summary: ChangeSummary;
  error?: string;
}

interface ParsedFunc {
  name: string;
  params: Param[];
  returns: string[];
  isAsync: boolean;
  isExported: boolean;
  startLine: number;
  endLine: number;
  bodyText: string;
}

interface ParsedType {
  name: string;
  kind: string;
  fields: Map<string, string>;
  isExported: boolean;
  startLine: number;
  endLine: number;
}

interface ParsedFile {
  functions: Map<string, ParsedFunc>;
  types: Map<string, ParsedType>;
  imports: Map<string, string>;
}

function parseFile(filePath: string): ParsedFile {
  const result: ParsedFile = {
    functions: new Map(),
    types: new Map(),
    imports: new Map(),
  };

  if (!filePath || !fs.existsSync(filePath)) {
    return result;
  }

  const content = fs.readFileSync(filePath, 'utf-8');
  const sourceFile = ts.createSourceFile(
    filePath,
    content,
    ts.ScriptTarget.Latest,
    true
  );

  function getLineNumber(pos: number): number {
    return sourceFile.getLineAndCharacterOfPosition(pos).line + 1;
  }

  function typeToString(type: ts.TypeNode | undefined): string {
    if (!type) return 'any';
    return content.substring(type.pos, type.end).trim();
  }

  function isExported(node: ts.Node): boolean {
    return (
      ts.canHaveModifiers(node) &&
      ts.getModifiers(node)?.some(
        (m) => m.kind === ts.SyntaxKind.ExportKeyword
      ) || false
    );
  }

  function visit(node: ts.Node) {
    // Extract imports
    if (ts.isImportDeclaration(node)) {
      const moduleSpecifier = node.moduleSpecifier;
      if (ts.isStringLiteral(moduleSpecifier)) {
        const importPath = moduleSpecifier.text;
        let alias = '';
        if (node.importClause?.name) {
          alias = node.importClause.name.text;
        }
        result.imports.set(importPath, alias);
      }
    }

    // Extract functions
    if (ts.isFunctionDeclaration(node) && node.name) {
      const func: ParsedFunc = {
        name: node.name.text,
        params: [],
        returns: [],
        isAsync: node.modifiers?.some(
          (m) => m.kind === ts.SyntaxKind.AsyncKeyword
        ) || false,
        isExported: isExported(node),
        startLine: getLineNumber(node.pos),
        endLine: getLineNumber(node.end),
        bodyText: node.body ? content.substring(node.body.pos, node.body.end) : '',
      };

      node.parameters.forEach((param) => {
        func.params.push({
          name: param.name.getText(sourceFile),
          type: typeToString(param.type),
        });
      });

      if (node.type) {
        func.returns.push(typeToString(node.type));
      }

      result.functions.set(func.name, func);
    }

    // Extract arrow functions assigned to const
    if (ts.isVariableStatement(node)) {
      const exported = isExported(node);
      node.declarationList.declarations.forEach((decl) => {
        if (
          ts.isIdentifier(decl.name) &&
          decl.initializer &&
          ts.isArrowFunction(decl.initializer)
        ) {
          const arrow = decl.initializer;
          const func: ParsedFunc = {
            name: decl.name.text,
            params: [],
            returns: [],
            isAsync: arrow.modifiers?.some(
              (m) => m.kind === ts.SyntaxKind.AsyncKeyword
            ) || false,
            isExported: exported,
            startLine: getLineNumber(node.pos),
            endLine: getLineNumber(node.end),
            bodyText: content.substring(arrow.body.pos, arrow.body.end),
          };

          arrow.parameters.forEach((param) => {
            func.params.push({
              name: param.name.getText(sourceFile),
              type: typeToString(param.type),
            });
          });

          if (arrow.type) {
            func.returns.push(typeToString(arrow.type));
          }

          result.functions.set(func.name, func);
        }
      });
    }

    // Extract interfaces
    if (ts.isInterfaceDeclaration(node)) {
      const parsedType: ParsedType = {
        name: node.name.text,
        kind: 'interface',
        fields: new Map(),
        isExported: isExported(node),
        startLine: getLineNumber(node.pos),
        endLine: getLineNumber(node.end),
      };

      node.members.forEach((member) => {
        if (ts.isPropertySignature(member) && member.name) {
          const name = member.name.getText(sourceFile);
          const type = typeToString(member.type);
          parsedType.fields.set(name, type);
        }
      });

      result.types.set(parsedType.name, parsedType);
    }

    // Extract type aliases
    if (ts.isTypeAliasDeclaration(node)) {
      const parsedType: ParsedType = {
        name: node.name.text,
        kind: 'type',
        fields: new Map(),
        isExported: isExported(node),
        startLine: getLineNumber(node.pos),
        endLine: getLineNumber(node.end),
      };

      if (ts.isTypeLiteralNode(node.type)) {
        node.type.members.forEach((member) => {
          if (ts.isPropertySignature(member) && member.name) {
            const name = member.name.getText(sourceFile);
            const type = typeToString(member.type);
            parsedType.fields.set(name, type);
          }
        });
      }

      result.types.set(parsedType.name, parsedType);
    }

    // Extract classes
    if (ts.isClassDeclaration(node) && node.name) {
      const parsedType: ParsedType = {
        name: node.name.text,
        kind: 'class',
        fields: new Map(),
        isExported: isExported(node),
        startLine: getLineNumber(node.pos),
        endLine: getLineNumber(node.end),
      };

      node.members.forEach((member) => {
        if (ts.isPropertyDeclaration(member) && member.name) {
          const name = member.name.getText(sourceFile);
          const type = typeToString(member.type);
          parsedType.fields.set(name, type);
        }
        // Extract class methods as functions
        if (ts.isMethodDeclaration(member) && member.name) {
          const methodName = `${node.name!.text}.${member.name.getText(sourceFile)}`;
          const func: ParsedFunc = {
            name: methodName,
            params: [],
            returns: [],
            isAsync: member.modifiers?.some(
              (m) => m.kind === ts.SyntaxKind.AsyncKeyword
            ) || false,
            isExported: isExported(node),
            startLine: getLineNumber(member.pos),
            endLine: getLineNumber(member.end),
            bodyText: member.body ? content.substring(member.body.pos, member.body.end) : '',
          };

          member.parameters.forEach((param) => {
            func.params.push({
              name: param.name.getText(sourceFile),
              type: typeToString(param.type),
            });
          });

          if (member.type) {
            func.returns.push(typeToString(member.type));
          }

          result.functions.set(func.name, func);
        }
      });

      result.types.set(parsedType.name, parsedType);
    }

    ts.forEachChild(node, visit);
  }

  visit(sourceFile);
  return result;
}

function compareFunctions(
  before: Map<string, ParsedFunc>,
  after: Map<string, ParsedFunc>
): FunctionDiff[] {
  const diffs: FunctionDiff[] = [];

  // Find removed and modified
  before.forEach((beforeFunc, name) => {
    const afterFunc = after.get(name);
    if (!afterFunc) {
      diffs.push({
        name,
        change_type: 'removed',
        before: {
          params: beforeFunc.params,
          returns: beforeFunc.returns,
          is_async: beforeFunc.isAsync,
          is_exported: beforeFunc.isExported,
          start_line: beforeFunc.startLine,
          end_line: beforeFunc.endLine,
        },
      });
      return;
    }

    // Check for modifications
    const sigChanged =
      JSON.stringify(beforeFunc.params) !== JSON.stringify(afterFunc.params) ||
      JSON.stringify(beforeFunc.returns) !== JSON.stringify(afterFunc.returns) ||
      beforeFunc.isAsync !== afterFunc.isAsync;

    const bodyChanged = beforeFunc.bodyText !== afterFunc.bodyText;

    if (sigChanged || bodyChanged) {
      const changes: string[] = [];
      if (JSON.stringify(beforeFunc.params) !== JSON.stringify(afterFunc.params)) {
        changes.push('parameters changed');
      }
      if (JSON.stringify(beforeFunc.returns) !== JSON.stringify(afterFunc.returns)) {
        changes.push('return type changed');
      }
      if (beforeFunc.isAsync !== afterFunc.isAsync) {
        changes.push('async modifier changed');
      }
      if (bodyChanged) {
        changes.push('implementation changed');
      }

      diffs.push({
        name,
        change_type: 'modified',
        before: {
          params: beforeFunc.params,
          returns: beforeFunc.returns,
          is_async: beforeFunc.isAsync,
          is_exported: beforeFunc.isExported,
          start_line: beforeFunc.startLine,
          end_line: beforeFunc.endLine,
        },
        after: {
          params: afterFunc.params,
          returns: afterFunc.returns,
          is_async: afterFunc.isAsync,
          is_exported: afterFunc.isExported,
          start_line: afterFunc.startLine,
          end_line: afterFunc.endLine,
        },
        body_diff: changes.join(', '),
      });
    }
  });

  // Find added
  after.forEach((afterFunc, name) => {
    if (!before.has(name)) {
      diffs.push({
        name,
        change_type: 'added',
        after: {
          params: afterFunc.params,
          returns: afterFunc.returns,
          is_async: afterFunc.isAsync,
          is_exported: afterFunc.isExported,
          start_line: afterFunc.startLine,
          end_line: afterFunc.endLine,
        },
      });
    }
  });

  return diffs;
}

function compareTypes(
  before: Map<string, ParsedType>,
  after: Map<string, ParsedType>
): TypeDiff[] {
  const diffs: TypeDiff[] = [];

  before.forEach((beforeType, name) => {
    const afterType = after.get(name);
    if (!afterType) {
      diffs.push({
        name,
        kind: beforeType.kind,
        change_type: 'removed',
        start_line: beforeType.startLine,
        end_line: beforeType.endLine,
      });
      return;
    }

    // Compare fields
    const fieldDiffs: FieldDiff[] = [];
    beforeType.fields.forEach((type, fieldName) => {
      const afterFieldType = afterType.fields.get(fieldName);
      if (!afterFieldType) {
        fieldDiffs.push({
          name: fieldName,
          change_type: 'removed',
          old_type: type,
        });
      } else if (afterFieldType !== type) {
        fieldDiffs.push({
          name: fieldName,
          change_type: 'modified',
          old_type: type,
          new_type: afterFieldType,
        });
      }
    });

    afterType.fields.forEach((type, fieldName) => {
      if (!beforeType.fields.has(fieldName)) {
        fieldDiffs.push({
          name: fieldName,
          change_type: 'added',
          new_type: type,
        });
      }
    });

    if (fieldDiffs.length > 0 || beforeType.kind !== afterType.kind) {
      diffs.push({
        name,
        kind: afterType.kind,
        change_type: 'modified',
        fields: fieldDiffs,
        start_line: afterType.startLine,
        end_line: afterType.endLine,
      });
    }
  });

  after.forEach((afterType, name) => {
    if (!before.has(name)) {
      diffs.push({
        name,
        kind: afterType.kind,
        change_type: 'added',
        start_line: afterType.startLine,
        end_line: afterType.endLine,
      });
    }
  });

  return diffs;
}

function compareImports(
  before: Map<string, string>,
  after: Map<string, string>
): ImportDiff[] {
  const diffs: ImportDiff[] = [];

  before.forEach((alias, importPath) => {
    if (!after.has(importPath)) {
      diffs.push({
        path: importPath,
        alias: alias || undefined,
        change_type: 'removed',
      });
    }
  });

  after.forEach((alias, importPath) => {
    if (!before.has(importPath)) {
      diffs.push({
        path: importPath,
        alias: alias || undefined,
        change_type: 'added',
      });
    }
  });

  return diffs;
}

function extractDiff(beforePath: string, afterPath: string): SemanticDiff {
  const before = parseFile(beforePath);
  const after = parseFile(afterPath);

  const functions = compareFunctions(before.functions, after.functions);
  const types = compareTypes(before.types, after.types);
  const imports = compareImports(before.imports, after.imports);

  const summary: ChangeSummary = {
    functions_added: functions.filter((f) => f.change_type === 'added').length,
    functions_removed: functions.filter((f) => f.change_type === 'removed').length,
    functions_modified: functions.filter((f) => f.change_type === 'modified').length,
    types_added: types.filter((t) => t.change_type === 'added').length,
    types_removed: types.filter((t) => t.change_type === 'removed').length,
    types_modified: types.filter((t) => t.change_type === 'modified').length,
    imports_added: imports.filter((i) => i.change_type === 'added').length,
    imports_removed: imports.filter((i) => i.change_type === 'removed').length,
  };

  return {
    language: 'typescript',
    file_path: afterPath || beforePath,
    functions,
    types,
    imports,
    summary,
  };
}

// CLI argument parsing
function parseArgs(args: string[]): { before: string; after: string } {
  let before = '';
  let after = '';

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--before' && i + 1 < args.length) {
      before = args[i + 1];
      i++;
    } else if (arg === '--after' && i + 1 < args.length) {
      after = args[i + 1];
      i++;
    } else if (!arg.startsWith('--')) {
      // Positional arguments for backward compatibility
      // Go wrapper calls: node script.js <before> <after>
      if (!before) {
        before = arg;
      } else if (!after) {
        after = arg;
      }
    }
  }

  // Handle empty string markers (Go wrapper sends '""' for empty paths)
  if (before === '""' || before === "''") {
    before = '';
  }
  if (after === '""' || after === "''") {
    after = '';
  }

  return { before, after };
}

// CLI entry point
function main() {
  const args = process.argv.slice(2);

  if (args.length < 2) {
    console.error('Usage: ast-extractor --before <before-path> --after <after-path>');
    console.error('Use empty string "" for new/deleted files');
    process.exit(1);
  }

  const { before: beforePath, after: afterPath } = parseArgs(args);

  if (!beforePath && !afterPath) {
    console.error('Error: At least one of --before or --after must be specified');
    process.exit(1);
  }

  try {
    const diff = extractDiff(beforePath, afterPath);
    console.log(JSON.stringify(diff, null, 2));
  } catch (error) {
    const diff: SemanticDiff = {
      language: 'typescript',
      file_path: afterPath || beforePath,
      functions: [],
      types: [],
      imports: [],
      summary: {
        functions_added: 0,
        functions_removed: 0,
        functions_modified: 0,
        types_added: 0,
        types_removed: 0,
        types_modified: 0,
        imports_added: 0,
        imports_removed: 0,
      },
      error: error instanceof Error ? error.message : String(error),
    };
    console.log(JSON.stringify(diff, null, 2));
    process.exit(1);
  }
}

main();

File: scripts/codereview/py/ast_extractor.py
#!/usr/bin/env python3
"""
Python AST Extractor for Semantic Diffs.

Extracts functions, classes, and imports from Python files
and compares before/after versions to generate semantic diffs.
"""

import ast
import hashlib
import json
import sys
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Optional


@dataclass
class Param:
    name: str
    type: str = ""


@dataclass
class FuncSig:
    params: list[Param]
    returns: list[str]
    is_async: bool = False
    decorators: list[str] = field(default_factory=list)
    is_exported: bool = True
    start_line: int = 0
    end_line: int = 0


@dataclass
class FunctionDiff:
    name: str
    change_type: str  # added, removed, modified, renamed
    before: Optional[FuncSig] = None
    after: Optional[FuncSig] = None
    body_diff: str = ""


@dataclass
class FieldDiff:
    name: str
    change_type: str
    old_type: str = ""
    new_type: str = ""


@dataclass
class TypeDiff:
    name: str
    kind: str  # class, dataclass
    change_type: str
    fields: list[FieldDiff] = field(default_factory=list)
    start_line: int = 0
    end_line: int = 0


@dataclass
class ImportDiff:
    path: str
    alias: str = ""
    change_type: str = ""


@dataclass
class ChangeSummary:
    functions_added: int = 0
    functions_removed: int = 0
    functions_modified: int = 0
    types_added: int = 0
    types_removed: int = 0
    types_modified: int = 0
    variables_added: int = 0
    variables_removed: int = 0
    variables_modified: int = 0
    imports_added: int = 0
    imports_removed: int = 0


@dataclass
class SemanticDiff:
    language: str
    file_path: str
    functions: list[FunctionDiff]
    types: list[TypeDiff]
    imports: list[ImportDiff]
    summary: ChangeSummary
    error: str = ""


@dataclass
class ParsedFunc:
    name: str
    params: list[Param]
    returns: list[str]
    is_async: bool
    decorators: list[str]
    is_exported: bool
    start_line: int
    end_line: int
    body_hash: str


@dataclass
class ParsedClass:
    name: str
    is_dataclass: bool
    fields: dict[str, str]  # name -> type
    methods: list[str]
    is_exported: bool
    start_line: int
    end_line: int


@dataclass
class ParsedFile:
    functions: dict[str, ParsedFunc]
    classes: dict[str, ParsedClass]
    imports: dict[str, str]  # module -> alias
    error: str = ""


def get_annotation_str(node: Optional[ast.expr]) -> str:
    """Convert an annotation AST node to string."""
    if node is None:
        return ""
    return ast.unparse(node)


def parse_file(file_path: str) -> ParsedFile:
    """Parse a Python file and extract semantic information."""
    result = ParsedFile(functions={}, classes={}, imports={})

    if not file_path or not Path(file_path).exists():
        return result

    content = Path(file_path).read_text()
    try:
        tree = ast.parse(content)
    except SyntaxError as e:
        result.error = f"Syntax error: {e.msg} at line {e.lineno}"
        return result

    for node in ast.walk(tree):
        # Extract imports
        if isinstance(node, ast.Import):
            for alias in node.names:
                result.imports[alias.name] = alias.asname or ""

        elif isinstance(node, ast.ImportFrom):
            module = node.module or ""
            for alias in node.names:
                key = f"{module}.{alias.name}" if module else alias.name
                result.imports[key] = alias.asname or ""

    # Process top-level definitions
    for node in ast.iter_child_nodes(tree):
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            func = _parse_function(node, content)
            result.functions[func.name] = func

        elif isinstance(node, ast.ClassDef):
            cls = _parse_class(node, content)
            result.classes[cls.name] = cls

            # Extract methods as functions
            for item in node.body:
                if isinstance(item, (ast.FunctionDef, ast.AsyncFunctionDef)):
                    method = _parse_function(item, content)
                    method.name = f"{cls.name}.{method.name}"
                    result.functions[method.name] = method

    return result


def _parse_function(
    node: ast.FunctionDef | ast.AsyncFunctionDef, content: str
) -> ParsedFunc:
    """Parse a function definition."""
    params = []
    for arg in node.args.args:
        params.append(Param(name=arg.arg, type=get_annotation_str(arg.annotation)))

    returns = []
    if node.returns:
        returns.append(get_annotation_str(node.returns))

    decorators = []
    for dec in node.decorator_list:
        if isinstance(dec, ast.Name):
            decorators.append(dec.id)
        elif isinstance(dec, ast.Call) and isinstance(dec.func, ast.Name):
            decorators.append(dec.func.id)
        elif isinstance(dec, ast.Attribute):
            decorators.append(ast.unparse(dec))

    # Hash the body for change detection
    end_line = node.end_lineno if node.end_lineno else node.lineno
    body_lines = content.split("\n")[node.lineno - 1 : end_line]
    body_hash = hashlib.sha256("\n".join(body_lines).encode("utf-8")).hexdigest()

    return ParsedFunc(
        name=node.name,
        params=params,
        returns=returns,
        is_async=isinstance(node, ast.AsyncFunctionDef),
        decorators=decorators,
        is_exported=not node.name.startswith("_"),
        start_line=node.lineno,
        end_line=node.end_lineno or node.lineno,
        body_hash=body_hash,
    )


def _parse_class(node: ast.ClassDef, content: str) -> ParsedClass:
    """Parse a class definition."""
    is_dataclass = any(
        (isinstance(d, ast.Name) and d.id == "dataclass")
        or (
            isinstance(d, ast.Call)
            and isinstance(d.func, ast.Name)
            and d.func.id == "dataclass"
        )
        for d in node.decorator_list
    )

    fields: dict[str, str] = {}
    methods: list[str] = []

    for item in node.body:
        if isinstance(item, ast.AnnAssign) and isinstance(item.target, ast.Name):
            fields[item.target.id] = get_annotation_str(item.annotation)
        elif isinstance(item, (ast.FunctionDef, ast.AsyncFunctionDef)):
            methods.append(item.name)

    return ParsedClass(
        name=node.name,
        is_dataclass=is_dataclass,
        fields=fields,
        methods=methods,
        is_exported=not node.name.startswith("_"),
        start_line=node.lineno,
        end_line=node.end_lineno or node.lineno,
    )


def compare_functions(
    before: dict[str, ParsedFunc], after: dict[str, ParsedFunc]
) -> list[FunctionDiff]:
    """Compare functions between before and after versions."""
    diffs = []

    # Find removed and modified
    for name, before_func in before.items():
        after_func = after.get(name)
        if after_func is None:
            diffs.append(
                FunctionDiff(
                    name=name,
                    change_type="removed",
                    before=FuncSig(
                        params=before_func.params,
                        returns=before_func.returns,
                        is_async=before_func.is_async,
                        decorators=before_func.decorators,
                        is_exported=before_func.is_exported,
                        start_line=before_func.start_line,
                        end_line=before_func.end_line,
                    ),
                )
            )
            continue

        # Check for modifications
        changes = []
        if before_func.params != after_func.params:
            changes.append("parameters changed")
        if before_func.returns != after_func.returns:
            changes.append("return type changed")
        if before_func.is_async != after_func.is_async:
            changes.append("async modifier changed")
        if before_func.decorators != after_func.decorators:
            changes.append("decorators changed")
        if before_func.body_hash != after_func.body_hash:
            changes.append("implementation changed")

        if changes:
            diffs.append(
                FunctionDiff(
                    name=name,
                    change_type="modified",
                    before=FuncSig(
                        params=before_func.params,
                        returns=before_func.returns,
                        is_async=before_func.is_async,
                        decorators=before_func.decorators,
                        is_exported=before_func.is_exported,
                        start_line=before_func.start_line,
                        end_line=before_func.end_line,
                    ),
                    after=FuncSig(
                        params=after_func.params,
                        returns=after_func.returns,
                        is_async=after_func.is_async,
                        decorators=after_func.decorators,
                        is_exported=after_func.is_exported,
                        start_line=after_func.start_line,
                        end_line=after_func.end_line,
                    ),
                    body_diff=", ".join(changes),
                )
            )

    # Find added
    for name, after_func in after.items():
        if name not in before:
            diffs.append(
                FunctionDiff(
                    name=name,
                    change_type="added",
                    after=FuncSig(
                        params=after_func.params,
                        returns=after_func.returns,
                        is_async=after_func.is_async,
                        decorators=after_func.decorators,
                        is_exported=after_func.is_exported,
                        start_line=after_func.start_line,
                        end_line=after_func.end_line,
                    ),
                )
            )

    return diffs


def compare_classes(
    before: dict[str, ParsedClass], after: dict[str, ParsedClass]
) -> list[TypeDiff]:
    """Compare classes between before and after versions."""
    diffs = []

    for name, before_cls in before.items():
        after_cls = after.get(name)
        if after_cls is None:
            diffs.append(
                TypeDiff(
                    name=name,
                    kind="dataclass" if before_cls.is_dataclass else "class",
                    change_type="removed",
                    start_line=before_cls.start_line,
                    end_line=before_cls.end_line,
                )
            )
            continue

        # Compare fields
        field_diffs = []
        for field_name, field_type in before_cls.fields.items():
            after_type = after_cls.fields.get(field_name)
            if after_type is None:
                field_diffs.append(
                    FieldDiff(
                        name=field_name,
                        change_type="removed",
                        old_type=field_type,
                    )
                )
            elif after_type != field_type:
                field_diffs.append(
                    FieldDiff(
                        name=field_name,
                        change_type="modified",
                        old_type=field_type,
                        new_type=after_type,
                    )
                )

        for field_name, field_type in after_cls.fields.items():
            if field_name not in before_cls.fields:
                field_diffs.append(
                    FieldDiff(
                        name=field_name,
                        change_type="added",
                        new_type=field_type,
                    )
                )

        if field_diffs or before_cls.is_dataclass != after_cls.is_dataclass:
            diffs.append(
                TypeDiff(
                    name=name,
                    kind="dataclass" if after_cls.is_dataclass else "class",
                    change_type="modified",
                    fields=field_diffs,
                    start_line=after_cls.start_line,
                    end_line=after_cls.end_line,
                )
            )

    for name, after_cls in after.items():
        if name not in before:
            diffs.append(
                TypeDiff(
                    name=name,
                    kind="dataclass" if after_cls.is_dataclass else "class",
                    change_type="added",
                    start_line=after_cls.start_line,
                    end_line=after_cls.end_line,
                )
            )

    return diffs


def compare_imports(before: dict[str, str], after: dict[str, str]) -> list[ImportDiff]:
    """Compare imports between before and after versions."""
    diffs = []

    for path, alias in before.items():
        if path not in after:
            diffs.append(ImportDiff(path=path, alias=alias, change_type="removed"))

    for path, alias in after.items():
        if path not in before:
            diffs.append(ImportDiff(path=path, alias=alias, change_type="added"))

    return diffs


def extract_diff(before_path: str, after_path: str) -> SemanticDiff:
    """Extract semantic diff between two Python files."""
    before = parse_file(before_path)
    after = parse_file(after_path)

    functions = compare_functions(before.functions, after.functions)
    types = compare_classes(before.classes, after.classes)
    imports = compare_imports(before.imports, after.imports)

    summary = ChangeSummary(
        functions_added=sum(1 for f in functions if f.change_type == "added"),
        functions_removed=sum(1 for f in functions if f.change_type == "removed"),
        functions_modified=sum(1 for f in functions if f.change_type == "modified"),
        types_added=sum(1 for t in types if t.change_type == "added"),
        types_removed=sum(1 for t in types if t.change_type == "removed"),
        types_modified=sum(1 for t in types if t.change_type == "modified"),
        imports_added=sum(1 for i in imports if i.change_type == "added"),
        imports_removed=sum(1 for i in imports if i.change_type == "removed"),
    )

    return SemanticDiff(
        language="python",
        file_path=after_path or before_path,
        functions=functions,
        types=types,
        imports=imports,
        summary=summary,
    )


def dataclass_to_dict(obj):
    """Recursively convert dataclass to dict, omitting empty values."""
    if hasattr(obj, "__dataclass_fields__"):
        result = {}
        for key, value in asdict(obj).items():
            if value is None:
                continue
            if isinstance(value, list) and not value:
                continue
            if isinstance(value, str) and not value:
                continue
            result[key] = value
        return result
    elif isinstance(obj, list):
        return [dataclass_to_dict(item) for item in obj]
    elif isinstance(obj, dict):
        return {k: dataclass_to_dict(v) for k, v in obj.items()}
    return obj


def main():
    """CLI entry point."""
    # Parse command line arguments
    before_path = ""
    after_path = ""

    i = 1
    while i < len(sys.argv):
        arg = sys.argv[i]
        if arg == "--before":
            if i + 1 < len(sys.argv):
                before_path = sys.argv[i + 1]
                i += 2
            else:
                print("Error: --before requires a path argument", file=sys.stderr)
                sys.exit(1)
        elif arg == "--after":
            if i + 1 < len(sys.argv):
                after_path = sys.argv[i + 1]
                i += 2
            else:
                print("Error: --after requires a path argument", file=sys.stderr)
                sys.exit(1)
        elif arg in ("--help", "-h"):
            print("Usage: ast_extractor.py --before <path> --after <path>")
            print()
            print("Options:")
            print("  --before <path>  Path to the before version of the file")
            print("  --after <path>   Path to the after version of the file")
            print()
            print('Use empty string "" for new/deleted files')
            sys.exit(0)
        else:
            # Positional arguments for backward compatibility
            if not before_path:
                before_path = arg
            elif not after_path:
                after_path = arg
            i += 1

    # Handle empty string markers
    if before_path in ('""', "''"):
        before_path = ""
    if after_path in ('""', "''"):
        after_path = ""

    if not before_path and not after_path:
        print("Usage: ast_extractor.py --before <path> --after <path>", file=sys.stderr)
        print('Use empty string "" for new/deleted files', file=sys.stderr)
        sys.exit(1)

    try:
        diff = extract_diff(before_path, after_path)
        output = dataclass_to_dict(diff)
        print(json.dumps(output, indent=2))
    except Exception as e:
        error_diff = SemanticDiff(
            language="python",
            file_path=after_path or before_path,
            functions=[],
            types=[],
            imports=[],
            summary=ChangeSummary(),
            error=str(e),
        )
        print(json.dumps(dataclass_to_dict(error_diff), indent=2))
        sys.exit(1)


if __name__ == "__main__":
    main()
